File: nodejs_implementation/src/chat/chat-interface.js
/**
 * Chat Interface for DeepWiki to Markdown Converter
 * 
 * This module provides the interface for interacting with the DeepWiki chat.
 */

const axios = require('axios');
const puppeteer = require('puppeteer');
const { logger } = require('../utils/logger');
const { handleRequestError, retryOnException } = require('../utils/error-handler');
const { convertHtmlToMarkdown } = require('../converters/html-to-md');

/**
 * Interface for interacting with the DeepWiki chat.
 * 
 * This class provides methods for sending messages to the DeepWiki chat
 * and receiving responses.
 */
class ChatInterface {
  /**
   * Initialize the chat interface.
   * 
   * @param {Object} config - Configuration object
   */
  constructor(config) {
    this.config = config;
    this.browser = null;
    this.page = null;
    this.session = axios.create({
      headers: config.headers || {},
      timeout: config.timeout || 30000
    });
    this.conversationId = null;
    
    logger.debug("Initialized ChatInterface");
  }
  
  /**
   * Connect to the DeepWiki chat.
   * 
   * @returns {Promise<boolean>} True if connection successful, False otherwise
   */
  async connect() {
    try {
      // Initialize Puppeteer if needed
      if (this.config.useSelenium) {
        await this._initPuppeteer();
      }
      
      // Create a new conversation
      await this._createConversation();
      
      logger.info("Connected to DeepWiki chat");
      return true;
    } catch (error) {
      logger.error(`Error connecting to DeepWiki chat: ${error.message}`);
      return false;
    }
  }
  
  /**
   * Disconnect from the DeepWiki chat.
   * 
   * @returns {Promise<void>}
   */
  async disconnect() {
    try {
      // Close Puppeteer browser if it was used
      if (this.browser) {
        await this.browser.close();
        this.browser = null;
        this.page = null;
      }
      
      logger.info("Disconnected from DeepWiki chat");
    } catch (error) {
      logger.error(`Error disconnecting from DeepWiki chat: ${error.message}`);
    }
  }
  
  /**
   * Send a message to the DeepWiki chat.
   * 
   * @param {string} message - Message to send
   * @returns {Promise<boolean>} True if message sent successfully, False otherwise
   */
  async sendMessage(message) {
    try {
      if (this.page) {
        return await this._sendMessagePuppeteer(message);
      } else {
        return await this._sendMessageApi(message);
      }
    } catch (error) {
      logger.error(`Error sending message: ${error.message}`);
      return false;
    }
  }
  
  /**
   * Get a response from the DeepWiki chat.
   * 
   * @param {number} [timeout=60000] - Maximum time to wait for a response in milliseconds
   * @returns {Promise<string|null>} Response text or null if no response received
   */
  async getResponse(timeout = 60000) {
    try {
      if (this.page) {
        return await this._getResponsePuppeteer(timeout);
      } else {
        return await this._getResponseApi(timeout);
      }
    } catch (error) {
      logger.error(`Error getting response: ${error.message}`);
      return null;
    }
  }
  
  /**
   * Initialize Puppeteer.
   * 
   * @returns {Promise<void>}
   * @private
   */
  async _initPuppeteer() {
    // Set up Puppeteer options
    const options = {
      headless: !this.config.showBrowser,
      args: ['--no-sandbox', '--disable-setuid-sandbox', '--disable-dev-shm-usage']
    };
    
    // Initialize browser
    this.browser = await puppeteer.launch(options);
    this.page = await this.browser.newPage();
    
    // Navigate to the chat URL
    await this.page.goto(this.config.url);
    
    // Wait for the page to load
    await this.page.waitForSelector('body');
    
    logger.debug("Initialized Puppeteer");
  }
  
  /**
   * Create a new conversation.
   * 
   * @returns {Promise<void>}
   * @private
   */
  async _createConversation() {
    if (this.page) {
      // Puppeteer-based conversation creation
      // This is a placeholder - actual implementation would depend on the DeepWiki UI
      return;
    } else {
      // API-based conversation creation
      const url = `${this.config.url}/api/conversation`;
      const data = {
        model: this.config.model || 'default',
        messages: []
      };
      
      try {
        const response = await this.session.post(url, data);
        
        if (response.data && response.data.conversation_id) {
          this.conversationId = response.data.conversation_id;
        } else {
          throw new Error("No conversation ID returned");
        }
      } catch (error) {
        handleRequestError(error, url);
        throw error;
      }
    }
  }
  
  /**
   * Send a message using Puppeteer.
   * 
   * @param {string} message - Message to send
   * @returns {Promise<boolean>} True if message sent successfully, False otherwise
   * @private
   */
  async _sendMessagePuppeteer(message) {
    try {
      // Find the input field
      const inputSelector = "textarea[placeholder*='message'], input[placeholder*='message']";
      await this.page.waitForSelector(inputSelector);
      
      // Clear the input field
      await this.page.evaluate((selector) => {
        document.querySelector(selector).value = '';
      }, inputSelector);
      
      // Type the message
      await this.page.type(inputSelector, message);
      
      // Find and click the send button
      const sendButtonSelector = "button[type='submit'], button.send-button";
      await this.page.waitForSelector(sendButtonSelector);
      await this.page.click(sendButtonSelector);
      
      // Wait for the message to be sent
      await this.page.waitForTimeout(1000);
      
      return true;
    } catch (error) {
      logger.error(`Error sending message with Puppeteer: ${error.message}`);
      return false;
    }
  }
  
  /**
   * Send a message using the API.
   * 
   * @param {string} message - Message to send
   * @returns {Promise<boolean>} True if message sent successfully, False otherwise
   * @private
   */
  _sendMessageApi = retryOnException(
    async (message) => {
      if (!this.conversationId) {
        throw new Error("No active conversation");
      }
      
      const url = `${this.config.url}/api/conversation/${this.conversationId}/message`;
      const data = {
        content: message,
        role: "user"
      };
      
      const response = await this.session.post(url, data);
      return true;
    },
    {
      maxRetries: this.config.retryCount || 3,
      retryDelay: this.config.retryDelay || 2000,
      retryOn: [axios.AxiosError]
    }
  );
  
  /**
   * Get a response using Puppeteer.
   * 
   * @param {number} timeout - Maximum time to wait for a response in milliseconds
   * @returns {Promise<string|null>} Response text or null if no response received
   * @private
   */
  async _getResponsePuppeteer(timeout) {
    try {
      // Wait for the response to appear
      const responseSelector = "div.assistant-message, div.response-content";
      
      // Wait for the response to stop changing (indicating it's complete)
      let lastResponse = null;
      const startTime = Date.now();
      
      while (Date.now() - startTime < timeout) {
        try {
          // Wait for the response element to be present
          await this.page.waitForSelector(responseSelector, { timeout: 5000 });
          
          // Get the current response text
          const currentResponse = await this.page.evaluate((selector) => {
            return document.querySelector(selector).innerHTML;
          }, responseSelector);
          
          // If the response hasn't changed for 2 seconds, consider it complete
          if (currentResponse === lastResponse) {
            // Convert HTML to Markdown
            return convertHtmlToMarkdown(currentResponse);
          }
          
          // Update last response
          lastResponse = currentResponse;
          
          // Wait a bit before checking again
          await this.page.waitForTimeout(2000);
        } catch (error) {
          // Response element not found yet, continue waiting
          await this.page.waitForTimeout(1000);
        }
      }
      
      // Timeout reached
      logger.warning(`Timeout reached while waiting for response`);
      return lastResponse ? convertHtmlToMarkdown(lastResponse) : null;
    } catch (error) {
      logger.error(`Error getting response with Puppeteer: ${error.message}`);
      return null;
    }
  }
  
  /**
   * Get a response using the API.
   * 
   * @param {number} timeout - Maximum time to wait for a response in milliseconds
   * @returns {Promise<string|null>} Response text or null if no response received
   * @private
   */
  _getResponseApi = retryOnException(
    async (timeout) => {
      if (!this.conversationId) {
        throw new Error("No active conversation");
      }
      
      const url = `${this.config.url}/api/conversation/${this.conversationId}/messages`;
      
      const startTime = Date.now();
      let lastMessageId = null;
      
      while (Date.now() - startTime < timeout) {
        const response = await this.session.get(url);
        const messages = response.data.messages || [];
        
        // Find the latest assistant message
        for (let i = messages.length - 1; i >= 0; i--) {
          const message = messages[i];
          if (message.role === 'assistant') {
            const messageId = message.id;
            
            // If this is a new message or the message has been updated
            if (messageId !== lastMessageId) {
              lastMessageId = messageId;
              
              // Check if the message is complete
              if (!message.is_incomplete) {
                return message.content || '';
              }
            }
            
            break;
          }
        }
        
        // Wait before checking again
        await new Promise(resolve => setTimeout(resolve, 2000));
      }
      
      // Timeout reached
      logger.warning(`Timeout reached while waiting for response`);
      return null;
    },
    {
      maxRetries: this.config.retryCount || 3,
      retryDelay: this.config.retryDelay || 2000,
      retryOn: [axios.AxiosError]
    }
  );
}

/**
 * Send a message to the DeepWiki chat.
 * 
 * @param {Object} config - Configuration object
 * @param {string} message - Message to send
 * @returns {Promise<boolean>} True if message sent successfully, False otherwise
 */
async function sendMessage(config, message) {
  const chat = new ChatInterface(config);
  
  try {
    await chat.connect();
    const result = await chat.sendMessage(message);
    return result;
  } finally {
    await chat.disconnect();
  }
}

/**
 * Send a message and get a response from the DeepWiki chat.
 * 
 * @param {Object} config - Configuration object
 * @param {string} message - Message to send
 * @param {number} [timeout=60000] - Maximum time to wait for a response in milliseconds
 * @returns {Promise<string|null>} Response text or null if no response received
 */
async function getResponse(config, message, timeout = 60000) {
  const chat = new ChatInterface(config);
  
  try {
    await chat.connect();
    
    if (await chat.sendMessage(message)) {
      return await chat.getResponse(timeout);
    }
    
    return null;
  } finally {
    await chat.disconnect();
  }
}

module.exports = {
  ChatInterface,
  sendMessage,
  getResponse
};

File: deepwiki_to_md/chat/chat_interface.py
"""
Chat Interface for DeepWiki to Markdown Converter

This module provides the interface for interacting with the DeepWiki chat.
"""

import logging
import time
import json
from typing import Dict, List, Optional, Any, Union
import requests
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.common.exceptions import TimeoutException, WebDriverException

from deepwiki_to_md.models.config import Config
from deepwiki_to_md.utils.error_handler import retry_on_exception, handle_request_error
from deepwiki_to_md.converters.html_to_md import convert_html_to_markdown

logger = logging.getLogger(__name__)


class ChatInterface:
    """
    Interface for interacting with the DeepWiki chat.

    This class provides methods for sending messages to the DeepWiki chat
    and receiving responses.
    """

    def __init__(self, config: Config):
        """
        Initialize the chat interface.

        Args:
            config (Config): Configuration object
        """
        self.config = config
        self.driver = None
        self.session = requests.Session()
        self.session.headers.update(config.headers)
        self.conversation_id = None

        logger.debug("Initialized ChatInterface")

    def __enter__(self):
        """
        Enter context manager.

        Returns:
            ChatInterface: Self
        """
        self.connect()
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        """
        Exit context manager.

        Args:
            exc_type: Exception type
            exc_val: Exception value
            exc_tb: Exception traceback
        """
        self.disconnect()

    def connect(self) -> bool:
        """
        Connect to the DeepWiki chat.

        Returns:
            bool: True if connection successful, False otherwise
        """
        try:
            # Initialize Selenium WebDriver if needed
            if hasattr(self.config, 'use_selenium') and self.config.use_selenium:
                self._init_webdriver()

            # Create a new conversation
            self._create_conversation()

            logger.info("Connected to DeepWiki chat")
            return True

        except Exception as e:
            logger.error(f"Error connecting to DeepWiki chat: {str(e)}")
            return False

    def disconnect(self) -> None:
        """
        Disconnect from the DeepWiki chat.
        """
        try:
            # Close Selenium WebDriver if it was used
            if self.driver:
                self.driver.quit()
                self.driver = None

            logger.info("Disconnected from DeepWiki chat")

        except Exception as e:
            logger.error(f"Error disconnecting from DeepWiki chat: {str(e)}")

    def send_message(self, message: str) -> bool:
        """
        Send a message to the DeepWiki chat.

        Args:
            message (str): Message to send

        Returns:
            bool: True if message sent successfully, False otherwise
        """
        try:
            if self.driver:
                return self._send_message_selenium(message)
            else:
                return self._send_message_api(message)

        except Exception as e:
            logger.error(f"Error sending message: {str(e)}")
            return False

    def get_response(self, timeout: int = 60) -> Optional[str]:
        """
        Get a response from the DeepWiki chat.

        Args:
            timeout (int): Maximum time to wait for a response in seconds

        Returns:
            Optional[str]: Response text or None if no response received
        """
        try:
            if self.driver:
                return self._get_response_selenium(timeout)
            else:
                return self._get_response_api(timeout)

        except Exception as e:
            logger.error(f"Error getting response: {str(e)}")
            return None

    def _init_webdriver(self) -> None:
        """
        Initialize Selenium WebDriver.
        """
        from webdriver_manager.chrome import ChromeDriverManager
        from selenium.webdriver.chrome.service import Service
        from selenium.webdriver.chrome.options import Options

        # Set up Chrome options
        chrome_options = Options()
        if not hasattr(self.config, 'show_browser') or not self.config.show_browser:
            chrome_options.add_argument('--headless')
        chrome_options.add_argument('--no-sandbox')
        chrome_options.add_argument('--disable-dev-shm-usage')

        # Initialize WebDriver
        service = Service(ChromeDriverManager().install())
        self.driver = webdriver.Chrome(service=service, options=chrome_options)

        # Navigate to the chat URL
        self.driver.get(self.config.url)

        # Wait for the page to load
        WebDriverWait(self.driver, 30).until(
            EC.presence_of_element_located((By.TAG_NAME, "body"))
        )

        logger.debug("Initialized Selenium WebDriver")

    def _create_conversation(self) -> None:
        """
        Create a new conversation.
        """
        if self.driver:
            # Selenium-based conversation creation
            # This is a placeholder - actual implementation would depend on the DeepWiki UI
            pass
        else:
            # API-based conversation creation
            url = f"{self.config.url}/api/conversation"
            data = {
                "model": getattr(self.config, 'library_name', 'default'),
                "messages": []
            }

            response = self.session.post(url=url, json=data)
            response.raise_for_status()

            result = response.json()
            self.conversation_id = result.get('conversation_id')

            if not self.conversation_id:
                raise ValueError("Failed to create conversation: No conversation ID returned")

    def _send_message_selenium(self, message: str) -> bool:
        """
        Send a message using Selenium.

        Args:
            message (str): Message to send

        Returns:
            bool: True if message sent successfully, False otherwise
        """
        try:
            # Find the input field
            input_field = WebDriverWait(self.driver, 10).until(
                EC.presence_of_element_located(
                    (By.CSS_SELECTOR, "textarea[placeholder*='message'], input[placeholder*='message']"))
            )

            # Clear the input field
            input_field.clear()

            # Type the message
            input_field.send_keys(message)

            # Find and click the send button
            send_button = WebDriverWait(self.driver, 10).until(
                EC.element_to_be_clickable((By.CSS_SELECTOR, "button[type='submit'], button.send-button"))
            )
            send_button.click()

            # Wait for the message to be sent
            time.sleep(1)

            return True

        except Exception as e:
            logger.error(f"Error sending message with Selenium: {str(e)}")
            return False

    @retry_on_exception
    def _send_message_api(self, message: str) -> bool:
        """
        Send a message using the API.

        Args:
            message (str): Message to send

        Returns:
            bool: True if message sent successfully, False otherwise
        """
        if not self.conversation_id:
            raise ValueError("No active conversation")

        url = f"{self.config.url}/api/conversation/{self.conversation_id}/message"
        data = {
            "content": message,
            "role": "user"
        }

        try:
            response = self.session.post(url, json=data)
            response.raise_for_status()
            return True

        except requests.RequestException as e:
            handle_request_error(e, url)
            raise

    def _get_response_selenium(self, timeout: int) -> Optional[str]:
        """
        Get a response using Selenium.

        Args:
            timeout (int): Maximum time to wait for a response in seconds

        Returns:
            Optional[str]: Response text or None if no response received
        """
        try:
            # Wait for the response to appear
            response_selector = "div.assistant-message, div.response-content"

            # Wait for the response to stop changing (indicating it's complete)
            last_response = None
            start_time = time.time()

            while time.time() - start_time < timeout:
                try:
                    # Wait for the response element to be present
                    response_element = WebDriverWait(self.driver, 5).until(
                        EC.presence_of_element_located((By.CSS_SELECTOR, response_selector))
                    )

                    # Get the current response text
                    current_response = response_element.get_attribute('innerHTML')

                    # If the response hasn't changed for 2 seconds, consider it complete
                    if current_response == last_response:
                        # Convert HTML to Markdown
                        return convert_html_to_markdown(current_response)

                    # Update last response
                    last_response = current_response

                    # Wait a bit before checking again
                    time.sleep(2)

                except TimeoutException:
                    # Response element not found yet, continue waiting
                    time.sleep(1)

            # Timeout reached
            logger.warning(f"Timeout reached while waiting for response")
            return last_response and convert_html_to_markdown(last_response)

        except Exception as e:
            logger.error(f"Error getting response with Selenium: {str(e)}")
            return None

    @retry_on_exception
    def _get_response_api(self, timeout: int) -> Optional[str]:
        """
        Get a response using the API.

        Args:
            timeout (int): Maximum time to wait for a response in seconds

        Returns:
            Optional[str]: Response text or None if no response received
        """
        if not self.conversation_id:
            raise ValueError("No active conversation")

        url = f"{self.config.url}/api/conversation/{self.conversation_id}/messages"

        start_time = time.time()
        last_message_id = None

        while time.time() - start_time < timeout:
            try:
                response = self.session.get(url)
                response.raise_for_status()

                messages = response.json().get('messages', [])

                # Find the latest assistant message
                for message in reversed(messages):
                    if message.get('role') == 'assistant':
                        message_id = message.get('id')

                        # If this is a new message or the message has been updated
                        if message_id != last_message_id:
                            last_message_id = message_id

                            # Check if the message is complete
                            if not message.get('is_incomplete', False):
                                return message.get('content', '')

                        break

                # Wait before checking again
                time.sleep(2)

            except requests.RequestException as e:
                handle_request_error(e, url)
                raise

        # Timeout reached
        logger.warning(f"Timeout reached while waiting for response")
        return None


def send_message(config: Config, message: str) -> bool:
    """
    Send a message to the DeepWiki chat.

    Args:
        config (Config): Configuration object
        message (str): Message to send

    Returns:
        bool: True if message sent successfully, False otherwise
    """
    with ChatInterface(config) as chat:
        return chat.send_message(message)


def get_response(config: Config, message: str, timeout: int = 60) -> Optional[str]:
    """
    Send a message and get a response from the DeepWiki chat.

    Args:
        config (Config): Configuration object
        message (str): Message to send
        timeout (int): Maximum time to wait for a response in seconds

    Returns:
        Optional[str]: Response text or None if no response received
    """
    with ChatInterface(config) as chat:
        if chat.send_message(message):
            return chat.get_response(timeout)
        return None
